cibuildwheel/linux.py

```diff
@@ -115,25 +129,72 @@ def build(project_dir, package_name, output_dir, test_command, test_requires, be
             gid=os.getgid(),
         )
 
-        docker_process = subprocess.Popen([
+        container_name = 'cibuildwheel-{}'.format(uuid.uuid4())
+
+        command = [
+            'docker',
+            'run',
+            '--env',
+            'CIBUILDWHEEL',
+            '--name', container_name,
+            '-i',
+            '-v', '/:/host', # ignored on Circle
+            docker_image,
+            '/bin/bash',
+        ]
+        print('docker command: {}'.format(command))
+        docker_process = subprocess.Popen(
+            command,
+            stdin=subprocess.PIPE,
+            universal_newlines=True,
+        )
+
+        # It will take a bit for docker to get the container up and
+        # running.  Keep trying to copy in the project directory until
+        # it succeeds.  There is a timeout to avoid spinning forever.
+
+        timeout = 180
+        start = monotonic.monotonic()
+        while True:
+            time.sleep(1)
+            command = [
                 'docker',
-                'run',
-                '--env',
-                'CIBUILDWHEEL',
-                '--rm',
-                '-i',
-                '-v', '%s:/project' % os.path.abspath(project_dir),
-                '-v', '%s:/output' % os.path.abspath(output_dir),
-                '-v', '/:/host',
-                docker_image,
-                '/bin/bash'],
-            stdin=subprocess.PIPE, universal_newlines=True)
+                'cp',
+                '{}/.'.format(os.path.abspath(project_dir)),
+                '{}:/project'.format(container_name),
+            ]
+            print('docker command: {}'.format(command))
+            if subprocess.call(command) == 0:
+                break
+
+            if monotonic.monotonic() - start > timeout:
+                raise DockerRunTimeoutError(
+                    'Unable to successfully copy project directory'
+                    ' within {} seconds'.format(timeout)
+                )
 
         try:
             docker_process.communicate(bash_script)
         except KeyboardInterrupt:
             docker_process.kill()
             docker_process.wait()
 
+        command = [
+            'docker',
+            'cp',
+            '{}:/output/.'.format(container_name),
+            os.path.abspath(output_dir),
+        ]
+        print('docker command: {}'.format(command))
+        subprocess.check_call(command)
+
+        command = [
```

Comment:
Would it be nice to have these different docker calls in separate functions, to keep the overview?

Reply:
(Or maybe collapse some of them to single lines?)

Reply:
Separate functions? Maybe.  For formatting, that's just what I do when it doesn't fit on one line.  Less judgement, more consistency, less diff noise when you change one thing.

Reply:
@YannickJadoul the PR is closed...  or I probably would have done some of this.  :]  Good ideas.

Reply:
Ah, I may have gotten a bit trigger happy in merging! Let's continue discussion in #96.